<h2 style="text-align:center">A Study on the Key Factors Affecting Singapore's <br>Purchasing Trend of HDB's Dwellings</h3>
<br>
<h3 style="text-align:center"> Tan Wei Xuan | DS102 Course Assignment (Sat 3:30PM)</h2>  

### Executive Summary:

***

### Introduction:

***

### Datasets:
<div class="alert alert-block alert-info">
<i> For Datasets 2 & 3, the data is adapted from their relevant stated links and reformatted to be usable for analysis. For Dataset 4, the data is created based on information extracted from Dataset 1's metadata </i>
</div>

1. `HDB Property Information` - [Data.gov.sg](https://data.gov.sg/dataset/hdb-property-information)
2. `Resident Households By Type Of Dwelling, Annual` - [Department of Statistics Singapore](https://www.tablebuilder.singstat.gov.sg/publicfacing/createDataTable.action?refId=12308)
3. `Average Household Size By Type of Dwelling` - [Department of Statistics Singaore](https://www.tablebuilder.singstat.gov.sg/publicfacing/createDataTable.action?refId=12305)
4. `Building Area Cluster Names` - *metadata-hdb-property-information (local file)*
5. `Planning Areas of Singapore` - [Wikitable (Web Scraping)](https://en.wikipedia.org/wiki/Planning_Areas_of_Singapore#List_of_Planning_Areas)
6. `Resale Flat Prices` - [Data.gov.sg](https://data.gov.sg/dataset/resale-flat-prices)


*** 

### Description of Datasets:
<div class="alert alert-block alert-info">
<i> Only the features of the datasets that are required/used for analysis are listed below. </i>
</div>

1. __`HDB Property Information`__describes **general statistics** on all the HDB Properties in Singapore according to the **location** of these Residential Units
<div style="margin-left:30px;margin-top:5px" >
    <ul style="list-style: none;background-color:#f0f1f2;">
        <li><b>blk_no:</b> <i>Block Number of the Residential Building</i></li>
        <li><b>street:</b> <i>Street Name of the Residential Building</i></li>
        <li><b>max_floor_lvl</b> <i>Maximum Floor Level</i></b></li>
        <li><b>year_completed:</b> <i>Year Completed</i></b></li>
        <li><b>residential:</b> <i>Residential Property Tag</i></li>
        <li><b>commercial:</b><i>Commercial Property Tag</i></li>
        <li><b>market_hawker:</b> <i>Market and Hawker Tag</i></li>
        <li><b>miscellaneous:</b> <i>Ameneties tag (Childcare, Committees Centre etc)</i></li>
        <li><b>multistorey_carpark:</b> <i>Multi-Storey Carpark Tag</i></li>
        <li><b>bldg_contract_town:</b> <i>Building Area Cluster</i></li>
        <li><b>total_dwelling_units:</b> <i>Total Dwelling Units Sold</i></li>
        <li><b>1room_sold:</b> <i>Number of 1-room sold flats</i></li>
        <li><b>2room_sold:</b> <i>Number of 2-room sold flats</i></li>
        <li><b>3room_sold:</b> <i>Number of 3-room sold flats</i></li>
        <li><b>4room_sold:</b> <i>Number of 4-room sold flats</i></li>
        <li><b>5room_sold:</b> <i>Number of 5-room sold flats</i></li>
        <li><b>exec_sold:</b> <i>Number of Executives sold flats</i></li>   
    </ul>
</div>
<br>

2. __`Resident Households By Type Of Dwelling, Annual`__ describes the annual total number of Resident Households based on the type of HDB Dwellings
<br>

3. __`Average Household Size By Type of Dwelling`__ describes the annual average household size based on the type of HDB Dwellings
<br>

4. __`Building Area Cluster Names`__ describes the **Names** and **Location** of **Building Area Clusters** according to their **respective prefixes**
<div style="margin-left:30px;margin-top:5px" >
    <ul style="list-style: none;background-color:#f0f1f2;">
        <li><b>Legend:</b> <i>Prefix of the Building Area Cluster</i></li>
        <li><b>Name:</b> <i>Name of the Building Area Cluster</i></li>
        <li><b>Latitude</b> <i>Latitude of the Building Area Cluster</i></li>
        <li><b>Longitude:</b> <i>Longitude of the Building Area Cluster</i></li>
    </ul>
</div>

<br>

5. __`Planning Areas of Singapore`__ describes the statistics on the **Building Area Clusters**
<div style="margin-left:30px;margin-top:5px" >
    <ul style="list-style: none;background-color:#f0f1f2;">
        <li><b>Name:</b> <i>Name of the Building Area Cluster</i></li>
        <li><b>Region:</b> <i>Region of the Building Area Cluster</i></li>
        <li><b>Area:</b><i>Area (km<sup>2</sup>) of the Building Area Cluster</i></li>
        <li><b>Population:</b><i>Population within the Building Area Cluster </i></li>
        <li><b>Density:</b> <i>Density (km<sup>2</sup>) of the Building Area Cluster</i></li>
    </ul>
</div>

*** 

### Dependencies:
<div class="alert alert-block alert-info">
Ensure that these libraries are installed on your local machine. The libraries <i>(1.Plotly w/ Cufflinks 2.Folium)</i> I used for plotting are not used/taught in class.
</div>

1. [__`Pandas`__](https://pandas.pydata.org/pandas-docs/stable/install.html)
2. [__`NumPy`__](https://scipy.org/install.html)
3. [__`Matplotlib`__](https://matplotlib.org/users/installing.html)
4. [__`Seaborn`__](https://seaborn.pydata.org/installing.html)
5. [__`Plotly`__](https://plot.ly/python/getting-started/)
6. [__`Cufflinks`__](https://pypi.org/project/cufflinks/)
7. [__`Folium`__](https://python-visualization.github.io/folium/installing.html)
8. [__`Requests`__](https://2.python-requests.org/en/master/user/install/)
9. [__`BeautifulSoup`__](https://pypi.org/project/beautifulsoup4/)
***

### Import Libaries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import folium as fol
from folium.plugins import HeatMap
import plotly.plotly as py
from plotly.offline import iplot, init_notebook_mode
from plotly.offline import download_plotlyjs, init_notebook_mode, plot,iplot
import plotly.graph_objs as go
import plotly.figure_factory as ff
import cufflinks as cf
import requests
from bs4 import BeautifulSoup

init_notebook_mode(connected=True)

# Display all floats rounded off to 2 decimal place
pd.options.display.float_format = '{:,.8f}'.format
# Plot static image in Jupyter notebook
%matplotlib inline 
# Settings throughout the notebook
sns.set()

<h2>Analysis of HDB Property Information</h3>

<h3>Cleaning of HDB Property's Dataset</h3>

We start off by reading the HDB Property Information.csv into a dataframe.

In [2]:
#Read the HDP Property Information into dataframes
hdb_prop_info_unfilt_df = pd.read_csv("hdb-property-information.csv")

hdb_prop_info_unfilt_df.head()

blk_no             street  max_floor_lvl  year_completed residential  \
0      1           BEACH RD             16            1970           Y   
1      1    BEDOK STH AVE 1             14            1975           Y   
2      1      CANTONMENT RD              2            2010           N   
3      1       CHAI CHEE RD             15            1982           Y   
4      1  CHANGI VILLAGE RD              4            1975           Y   

  commercial market_hawker miscellaneous multistorey_carpark  \
0          Y             N             N                   N   
1          N             N             Y                   N   
2          Y             N             N                   N   
3          N             N             N                   N   
4          Y             N             N                   N   

  precinct_pavilion  ... 3room_sold  4room_sold  5room_sold  exec_sold  \
0                 N  ...        138           1           2          0   
1                 N  ...        204           0           2          0   
2                 N  ...          0           0           0          0   
3                 N  ...          0          10          92          0   
4                 N  ...         54           0           1          0   

   multigen_sold  studio_apartment_sold  1room_rental  2room_rental  \
0              0                      0             0             0   
1              0                      0             0             0   
2              0                      0             0             0   
3              0                      0             0             0   
4              0                      0             0             0   

   3room_rental  other_room_rental  
0             0                  0  
1             0                  0  
2             0                  0  
3             0                  0  
4             0                  0  

[5 rows x 24 columns]

The dataframe consist of several columns that are __not required for analysis__. As not all of the features of the dataset are needed, we extract the __required features__ only.

In [38]:
#Extract the required features
hdb_prop_info_df =  hdb_prop_info_unfilt_df[['blk_no',
                                             'street',
                                             'max_floor_lvl',
                                             'year_completed',
                                             'residential',
                                             'commercial',
                                             'market_hawker',
                                             'miscellaneous',
                                             'multistorey_carpark',
                                             'bldg_contract_town',
                                             'total_dwelling_units',
                                             '1room_sold',
                                             '2room_sold',
                                             '3room_sold',
                                             '4room_sold',
                                             '5room_sold',
                                             'exec_sold',
                                             'multigen_sold']]

#Checking for empty fields
null_columns = hdb_prop_info_df.columns[hdb_prop_info_df.isnull().any()]
print("There are", len(null_columns), "empty fields")

# There are 0 empty fields and thus we do not have to drop any NaN rows.

hdb_prop_info_df.head()

There are 0 empty fields


blk_no             street  max_floor_lvl  year_completed residential  \
0      1           BEACH RD             16            1970           Y   
1      1    BEDOK STH AVE 1             14            1975           Y   
2      1      CANTONMENT RD              2            2010           N   
3      1       CHAI CHEE RD             15            1982           Y   
4      1  CHANGI VILLAGE RD              4            1975           Y   

  commercial market_hawker miscellaneous multistorey_carpark  \
0          Y             N             N                   N   
1          N             N             Y                   N   
2          Y             N             N                   N   
3          N             N             N                   N   
4          Y             N             N                   N   

  bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
0                KWN                   142           0           1   
1                 BD                   206           0           0   
2                 CT                     0           0           0   
3                 BD                   102           0           0   
4                PRC                    55           0           0   

   3room_sold  4room_sold  5room_sold  exec_sold  multigen_sold  
0         138           1           2          0              0  
1         204           0           2          0              0  
2           0           0           0          0              0  
3           0          10          92          0              0  
4          54           0           1          0              0

After extracting the required features, there are still several columns that needs to be cleaned. The columns are **residential**, **commercial**, **blk_no**,**street** and **bldg_contract_town**.

As we are doing an analysis for __Residential Units only__, we want to drop all the rows that are non-residential. 

In [39]:
hdb_prop_info_df = hdb_prop_info_df.drop(hdb_prop_info_df[hdb_prop_info_df.residential == 'N'].index)

hdb_prop_info_df.head()

blk_no             street  max_floor_lvl  year_completed residential  \
0      1           BEACH RD             16            1970           Y   
1      1    BEDOK STH AVE 1             14            1975           Y   
3      1       CHAI CHEE RD             15            1982           Y   
4      1  CHANGI VILLAGE RD              4            1975           Y   
5      1          DELTA AVE             25            1982           Y   

  commercial market_hawker miscellaneous multistorey_carpark  \
0          Y             N             N                   N   
1          N             N             Y                   N   
3          N             N             N                   N   
4          Y             N             N                   N   
5          N             N             N                   N   

  bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
0                KWN                   142           0           1   
1                 BD                   206           0           0   
3                 BD                   102           0           0   
4                PRC                    55           0           0   
5                 BM                    96           0           0   

   3room_sold  4room_sold  5room_sold  exec_sold  multigen_sold  
0         138           1           2          0              0  
1         204           0           2          0              0  
3           0          10          92          0              0  
4          54           0           1          0              0  
5           0           0          96          0              0

The dataset will thus be left with buildings that are either fully residential *(By Tag: Residential = Y ,Commercial = N)* or are both Residential and Commercial *(By Tag: Residential = Y, Commercial = Y)*

We concatenate the __blk_no__ column with the __street__ column to make it easier to identify/understand the location of the respective HDB residential units. We create a new column __'Address'__ to store the new values obtained.

In [40]:
hdb_prop_info_df['Address'] = hdb_prop_info_df[['blk_no', 'street']].apply(lambda x: ' '.join(x), axis=1)

# We reorder the columns to put the ADDRESS column in front.
cols = list(hdb_prop_info_df.columns)
cols = [cols[-1]] + cols[:-1]
hdb_prop_info_df = hdb_prop_info_df[cols]

# With the Address Column, we can now drop the blk_no & street column
hdb_prop_info_df = hdb_prop_info_df.drop(columns=['blk_no', 'street'], axis=1)

hdb_prop_info_df.head()

Address  max_floor_lvl  year_completed residential commercial  \
0           1 BEACH RD             16            1970           Y          Y   
1    1 BEDOK STH AVE 1             14            1975           Y          N   
3       1 CHAI CHEE RD             15            1982           Y          N   
4  1 CHANGI VILLAGE RD              4            1975           Y          Y   
5          1 DELTA AVE             25            1982           Y          N   

  market_hawker miscellaneous multistorey_carpark bldg_contract_town  \
0             N             N                   N                KWN   
1             N             Y                   N                 BD   
3             N             N                   N                 BD   
4             N             N                   N                PRC   
5             N             N                   N                 BM   

   total_dwelling_units  1room_sold  2room_sold  3room_sold  4room_sold  \
0                   142           0           1         138           1   
1                   206           0           0         204           0   
3                   102           0           0           0          10   
4                    55           0           0          54           0   
5                    96           0           0           0           0   

   5room_sold  exec_sold  multigen_sold  
0           2          0              0  
1           2          0              0  
3          92          0              0  
4           1          0              0  
5          96          0              0

The __bld_contract_town__ column is a prefix correlating to the Name of the Building Area Cluster and to make it  more comprehensible, we can replace the prefix with their relevant __Building Area Cluster names__

In [41]:
#Import the Building Area Culster Names data to a dataframe
bld_area_cluster_names_df = pd.read_csv("Building Area Cluster Names.csv")

In [42]:
#Create the function to match prefix -> name 
def prefix_to_name(prefix):
    for index, row in bld_area_cluster_names_df.iterrows():
        if(prefix == row[0]):
            return row[1]


In [43]:
#This will take some time to load as we are looping and replacing alot of values
hdb_prop_info_df['bldg_contract_town'] = hdb_prop_info_df['bldg_contract_town'].apply(prefix_to_name)

hdb_prop_info_df.head()

Address  max_floor_lvl  year_completed residential commercial  \
0           1 BEACH RD             16            1970           Y          Y   
1    1 BEDOK STH AVE 1             14            1975           Y          N   
3       1 CHAI CHEE RD             15            1982           Y          N   
4  1 CHANGI VILLAGE RD              4            1975           Y          Y   
5          1 DELTA AVE             25            1982           Y          N   

  market_hawker miscellaneous multistorey_carpark bldg_contract_town  \
0             N             N                   N            KALLANG   
1             N             Y                   N              BEDOK   
3             N             N                   N              BEDOK   
4             N             N                   N          PASIR RIS   
5             N             N                   N        BUKIT MERAH   

   total_dwelling_units  1room_sold  2room_sold  3room_sold  4room_sold  \
0                   142           0           1         138           1   
1                   206           0           0         204           0   
3                   102           0           0           0          10   
4                    55           0           0          54           0   
5                    96           0           0           0           0   

   5room_sold  exec_sold  multigen_sold  
0           2          0              0  
1           2          0              0  
3          92          0              0  
4           1          0              0  
5          96          0              0

<h2 style='text-decoration:underline'>Number of Residential Areas Completed by Year</h2>


We group the building area clusters with respect to the region that they are in. <br>
We use the mean of the latitude and longitude of the building clusters within the same region to find the most centralized location of that region.

In [81]:
# We use the mean of latitude and longitude to find the most centralized location of the region based on the building area clusters within the region
hdb_prop_region_gb = hdb_prop_full_added_df.groupby(['Region'], as_index=False).agg({"total_dwelling_units": "sum",
                                                                                     "latitude":"mean",
                                                                                     "longitude": "mean"})


In [148]:
red_area_by_year__gb = hdb_prop_info_df.groupby(['year_completed'], as_index=False).agg({"Address": "count"})

red_area_by_year__gb.head()

year_completed  Address
0            1937       20
1            1938        2
2            1949       30
3            1955        2
4            1958        8

In [143]:
trace0 = go.Scatter(
    x = red_area_by_year__gb['year_completed'],
    y = red_area_by_year__gb['Address'],
    mode = 'markers',
    marker=dict(
        size= red_area_by_year__gb['Address'],
        sizemode='area',
        sizeref=2.*max(red_area_by_year__gb['Address'])/(40.**2),
        sizemin=4,
        color = 'rgb(44, 160, 101)'
    )
)

data = [trace0]

layout = dict(title = 'Number of Residential Areas Completed by Year',
              yaxis = dict(zeroline = False, title = "Number of Residential Areas"),
              xaxis = dict(zeroline = False, title = "Year Completed")
             )

fig = dict(data=data, layout=layout)
iplot(fig, filename='bubblechart-size-ref')

<h3><u>ANALYSIS</u></h3>
sdfdsf

<h2 style='text-decoration:underline'>Dwelling Units Sold based on Building Year Completion of Residential Areas</h2>


We will analyze the trend of the number dwelling units by type being sold over the years

In [47]:
hdb_prop_year_gb = hdb_prop_info_df.groupby(['year_completed'], as_index=False).agg({"1room_sold":"sum",
                                                                                    "2room_sold":"sum",
                                                                                    "3room_sold":"sum",
                                                                                    "4room_sold":"sum",
                                                                                    "5room_sold":"sum",
                                                                                    "exec_sold":"sum",
                                                                                    "multigen_sold": "sum"})

Plot the **Line Graph** for the **amount of each dwelling types sold based on Building Year Completion of Residential Area**

In [122]:
trace0 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['1room_sold'],
                    name = '1-Room Sold',
                    opacity = 0.5
                   )

trace1 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['2room_sold'],
                    name = '2-Room Sold',
                    opacity = 0.5
                  )

trace2 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['3room_sold'],
                    name = '3-Room Sold',
                    opacity = 0.5
                  )


trace3 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['4room_sold'],
                    name = '4-Room Sold',
                    opacity = 0.5
                  )


trace4 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['5room_sold'],
                    name = '5-Room Sold',
                    opacity = 0.5
                  )


trace5 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['exec_sold'],
                    name = 'Executive Sold',
                    opacity = 0.5
                  )

trace6 = go.Scatter(
                    x = hdb_prop_year_gb['year_completed'] ,
                    y = hdb_prop_year_gb['multigen_sold'],
                    name = 'Multi-Generation Sold',
                    opacity = 0.5
                  )


data = [trace0, trace1, trace2,trace3, trace4, trace5, trace6]


layout = dict(
        
        title = "Total Dwelling Units Sold based on Building Year Completion of Residential Area",
        xaxis = dict(title ="Year Completed",automargin=True),
        yaxis = dict(title = "Total Dwelling Units Sold"),
        autosize=True
)


fig = dict(data=data, layout=layout)

iplot(fig)



<h3><u>ANALYSIS</u></h3>
sdfdsf

<h2 style='text-decoration:underline'>Distribution of Dwelling Units Sold</h2>

We will analyze the **distribution of dwelling units** for the entire dataset.

In [49]:
print("The skewness of the distribution of the dwelling units is: ", hdb_prop_info_df['total_dwelling_units'].skew())

data = [
  go.Histogram(
    x = hdb_prop_info_df['total_dwelling_units'],
    name = "count",
    xbins=dict(
        start=0.0,
        end= 600,
        size=15
    ),
    marker=dict(
        color='#EB89B5'
    ),
    opacity=0.75
  )
]


layout = go.Layout(
        title = 'Distribution of Dwelling units Sold',
        xaxis = dict(title = 'Dwelling Units'),
        yaxis = dict(title = 'Frequency'),
        shapes=[dict({
            'type': 'line',
            'x0': hdb_prop_info_df['total_dwelling_units'].mean(),
            'y0': 0,
            'x1': hdb_prop_info_df['total_dwelling_units'].mean(),
            'y1': 1800,
            'line': {
                'color': '#8bbe1b',
                'width': 3,
                'dash' : 'dashdot'
            }}),
            dict({
            'type': 'line',
            'x0': hdb_prop_info_df['total_dwelling_units'].median(),
            'y0': 0,
            'x1': hdb_prop_info_df['total_dwelling_units'].median(),
            'y1': 1800,
            'line': {
                'color': '#e12c2c',
                'width': 3,
                'dash' : 'dashdot'
            }}),
            dict({
            'type': 'line',
            'x0': hdb_prop_info_df['total_dwelling_units'].mode()[0],
            'y0': 0,
            'x1': hdb_prop_info_df['total_dwelling_units'].mode()[0],
            'y1': 1800,
            'line': {
                'color': '#cc338b',
                'width': 3,
                'dash' : 'dashdot'
            }})],
        annotations=[
        dict(
            x=500,
            y=1800,
            xref='x',
            yref='y',
            text="Skewness: " +  str(hdb_prop_info_df['total_dwelling_units'].skew()),
            showarrow=False,
            ax=0,
            ay=-40
        ),
        dict(
            x=500,
            y=1700,
            xref='x',
            yref='y',
            text="Mean: " +  str(hdb_prop_info_df['total_dwelling_units'].mean()),
            showarrow=False,
            ax=0,
            ay=-40
        ),
        dict(
            x=500,
            y=1600,
            xref='x',
            yref='y',
            text="Median: " +  str(hdb_prop_info_df['total_dwelling_units'].median()),
            showarrow=False,
            ax=0,
            ay=-40
        ),
        dict(
            x=500,
            y=1500,
            xref='x',
            yref='y',
            text="Mode: " +  str(hdb_prop_info_df['total_dwelling_units'].mode()[0]),
            showarrow=False,
            ax=0,
            ay=-40
        ),
       dict(
            x=hdb_prop_info_df['total_dwelling_units'].mean(),
            y=1800,
            xref='x',
            yref='y',
            text='Mean',
            showarrow=True,
            arrowhead=1,
            ax=0,
            ay=-70,
            arrowwidth = 1
        ),
       dict(
            x=hdb_prop_info_df['total_dwelling_units'].median(),
            y=1800,
            xref='x',
            yref='y',
            text='Median',
            showarrow=True,
            arrowhead=1,
            ax=0,
            ay=-50,
            arrowwidth = 1
        ),
      dict(
            x=hdb_prop_info_df['total_dwelling_units'].mode()[0],
            y=1800,
            xref='x',
            yref='y',
            text='Mode',
            showarrow=True,
            arrowhead=1,
            ax= 0,
            ay=-30,
            arrowwidth = 1
        ) 
    ]
)

fig = go.Figure(data=data, layout=layout)

iplot(fig, filename='histogram')

The skewness of the distribution of the dwelling units is:  1.6965754292951836


<h3><u>ANALYSIS</u></h3>

The Distribution of the Dwelling Units Sold for the dataset is **right-tailed** and has a **positive skewnewss** of **1.6965754292951836**. As the **skewness value is > 1**, the distribution is **highly skewed**. The mass of the distribution is **concentrated on the left of the figure**. All the locations in Singapore that have residential units in Singapore have an **average of 106.6 units sold**.

The **mean** of the Dwelling Units Sold is **greater than** the **median** and both of them are **greater** than the **mode**.

<h2 style='text-decoration:underline'>Percentage of Dwelling Units Sold By Type</h2> 

We will analyze the **percentage of dwelling units sold by type** for the entire dataset to find out
which is the **most common type of dwelling unit** type that is sold .

In [51]:
labels = ['1 Room','2 Room', '3 Room','4 Room', '5 Room', 'Executive', 'Multi-Generation']
values = [hdb_prop_info_df['1room_sold'].sum(),
          hdb_prop_info_df['2room_sold'].sum(),
          hdb_prop_info_df['3room_sold'].sum(),
          hdb_prop_info_df['4room_sold'].sum(),
          hdb_prop_info_df['5room_sold'].sum(),
          hdb_prop_info_df['exec_sold'].sum(),
          hdb_prop_info_df['multigen_sold'].sum()]

colors = ['#FEBFB3', '#E1396C', '#D0F9B1', '#96D38C', '#D000B1' ,'#9642f5','#8674f3']

trace = go.Pie(labels = labels, 
               values = values,
               hoverinfo ='label+percent', 
               textinfo = 'label+percent', 
               textfont = dict(size=13),
               title = dict(text ='Percentage of Dwelling Units Sold By Type',
                            font = dict(size = 16),
                            position = 'top left'
                           ),
               marker= dict(colors=colors, 
                            line=dict(color='#000000', 
                                      width=2)
                           )
              )
                    

iplot([trace], filename='styled_pie_chart')

<h3><u>ANALYSIS</u></h3>

<h2 style='text-decoration:underline'>Dwelling Units Sold based on Residential and Commercial Tag</h2>

We will analyze the trend of the number dwelling units based on whether the residential estate is ~**purely for residential purposes** or if the estate comprises of **both residential and commercial facilities**

We group the data according to their Residential and Commericial Tags and get the total number of dwelling units in these clusters

In [54]:
hdb_prop_residential_gb = hdb_prop_info_df.groupby(['residential','commercial'], as_index=False).agg({"total_dwelling_units": "sum",
                                                                                    "1room_sold":"sum",
                                                                                    "2room_sold":"sum",
                                                                                    "3room_sold":"sum",
                                                                                    "4room_sold":"sum",
                                                                                    "5room_sold":"sum",
                                                                                    "exec_sold":"sum",
                                                                                    "multigen_sold": "sum"})

hdb_prop_residential_gb.head()

residential commercial  total_dwelling_units  1room_sold  2room_sold  \
0           Y          N                863570         255       21386   
1           Y          Y                199536           0        1972   

   3room_sold  4room_sold  5room_sold  exec_sold  multigen_sold  
0      168604      344370      209732      58741            259  
1       72104       75699       31949       5996            108

In [59]:
room_type_name = {'Name':['1-Room', '2-Room', '3-Room', '4-Room', '5-Room','Executive','Multi-Generation']} 
  
room_type_name_df = pd.DataFrame(data)

trace1 = go.Bar(
    x= room_type_name['Name'],
    y= hdb_prop_residential_gb.iloc[0][3:10],
    name='Residential Only',
    textposition = 'auto',
    marker=dict(
        color='rgb(58,200,225)',
        line=dict(
            color='rgb(8,48,107)',
            width=1.5),
        ),
    opacity=0.6
)
trace2 = go.Bar(
    x= room_type_name['Name'],
    y= hdb_prop_residential_gb.iloc[1][3:10],
    name='Residential and Commercial',
    textposition = 'auto',
    marker=dict(
        color='rgb(158,202,225)',
        line=dict(
            color='rgb(8,48,107)',
            width=1.5),
        ),
    opacity=0.6
)

data = [trace1, trace2]

layout = go.Layout(
    barmode='group',
    title = 'Dwelling Units Sold by Type based on Residential and Commercial Tag',
    xaxis = dict(title='Type of Dwelling Units'),
    yaxis = dict(title='Number of Dwelling Units Sold') 
)

fig = go.Figure(data=data, layout=layout)

py.iplot(fig, filename='grouped-bar-direct-labels')



<h3><u>ANALYSIS</u></h3>
sdfsdfsdf

# Amenities Graph here

### Group Data for further analysis

We group the data according to their building cluster names and get the total number of dwelling units sold in these clusters

In [21]:
hdb_prop_gb = hdb_prop_info_df.groupby(['bldg_contract_town'], as_index=False).agg({"total_dwelling_units": "sum",
                                                                                    "1room_sold":"sum",
                                                                                    "2room_sold":"sum",
                                                                                    "3room_sold":"sum",
                                                                                    "4room_sold":"sum",
                                                                                    "5room_sold":"sum",
                                                                                    "exec_sold":"sum"})
hdb_prop_gb = hdb_prop_gb.sort_values(by=['total_dwelling_units'])

#Remove leading and trailing whitespaces
hdb_prop_gb['bldg_contract_town'] = hdb_prop_gb['bldg_contract_town'] .apply(lambda x : x.strip())
hdb_prop_gb.head()

bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
6         BUKIT TIMAH                  2554           0          49   
15      MARINE PARADE                  7860           0          26   
7        CENTRAL AREA                 12003           0         345   
2              BISHAN                 20072           0          17   
21          SERANGOON                 21634           0         161   

    3room_sold  4room_sold  5room_sold  exec_sold  
6          439         920         683        380  
15        3032        1798        1679          0  
7         4214        3424         902          6  
2         2357        9359        5716       1567  
21        4535       10231        3757       2365

### Web Scraping to get more information about each of the Building Area Clusters

Next, we want to scrape the **Planning Areas of Singapore's Data** from **Wikipedia** to obtain **more information** about the **respective Building Area Clusters**

Set up a connection to the server

In [22]:
url = "https://en.wikipedia.org/wiki/Planning_Areas_of_Singapore#List_of_Planning_Areas"

resp = requests.get(url)

planning_areas_wiki_soup = BeautifulSoup(resp.text, 'html.parser')

if(resp.status_code == 200):
    print("Connection is Sucessful (Status Code:", resp.status_code, ")") 
else:
    print("Connection is Unsuccesful (Status Code:", resp.status_code, ")") 

Connection is Sucessful (Status Code: 200 )


Get all the Columns and Rows of the Table that we want to scrape

In [23]:
planning_areas_table_tag = planning_areas_wiki_soup.find('table', attrs={'class':"wikitable sortable"})

all_planning_areas_rows = planning_areas_table_tag.find_all('tr')

planning_areas_col = planning_areas_table_tag.find_all('th')
planning_areas_headers = []

for hc in planning_areas_col:
        planning_areas_headers.append(hc.text.strip())
        

Store all data according to their rows and columns

In [24]:
planning_areas_dict = {}

indexer = 0

planning_areas_rows = all_planning_areas_rows[1:]

for header in planning_areas_headers:
    
    column_list = []
    
    for row in planning_areas_rows:
        
        td_tags = row.findAll('td')
        
        data_cell = td_tags[indexer].text.strip() # takes only the cell it wants based on the indexer
        column_list.append(data_cell)
    
    indexer += 1
    planning_areas_dict[header] = column_list

Store collected data into a dataframe

In [25]:
planning_areas_df = pd.DataFrame(planning_areas_dict)
planning_areas_df.head()

Name (English) Malay Chinese         Pinyin              Tamil      Region  \
0     Ang Mo Kio           宏茂桥  Hóng mào qiáo        ஆங் மோ கியோ  North-East   
1          Bedok     *      勿洛         Wù luò             பிடோக்        East   
2         Bishan            碧山        Bì shān             பீஷான்     Central   
3       Boon Lay            文礼         Wén lǐ            பூன் லே        West   
4    Bukit Batok     *    武吉巴督     Wǔjí bā dū  புக்கிட் பாத்தோக்        West   

  Area (km2) Population[6] Density (/km2)  
0      13.94       165,710         12,000  
1      21.69       281,300         13,000  
2       7.62        88,490         12,000  
3       8.23            30            3.6  
4      11.13       144,410         13,000

Clean the dataframe to hold only the columns we need

In [26]:
planning_areas_df = planning_areas_df[['Name (English)','Region' ,'Area (km2)', 'Population[6]', 'Density (/km2)']]
planning_areas_df = planning_areas_df.rename({'Name (English)':'bldg_contract_town', 'Area (km2)':'Area', 'Population[6]':'Population', 'Density (/km2)':'Density'}, axis='columns')

planning_areas_df.head(n=10)

bldg_contract_town      Region   Area Population Density
0               Ang Mo Kio  North-East  13.94    165,710  12,000
1                    Bedok        East  21.69    281,300  13,000
2                   Bishan     Central   7.62     88,490  12,000
3                 Boon Lay        West   8.23         30     3.6
4              Bukit Batok        West  11.13    144,410  13,000
5              Bukit Merah     Central  14.34    151,870  11,000
6            Bukit Panjang        West   8.99    140,820  16,000
7              Bukit Timah     Central  17.53     77,280   4,400
8  Central Water Catchment       North  37.15          *       *
9                   Changi        East  40.61      2,080    62.3

Rename the Columns and Replace Empty Rows (Have to do this as the dataset we scraped keep the values in empty cells as *) with NaN

In [27]:
planning_areas_df = planning_areas_df[["bldg_contract_town", "Region" ,"Area", "Population","Density"]].astype(str).replace(['*',''],np.nan)

planning_areas_df = planning_areas_df.dropna()

#Change the Cluster Name to Upper case to match with the other datasets
planning_areas_df['bldg_contract_town'] = planning_areas_df['bldg_contract_town'] .apply(lambda x : x.upper())

planning_areas_df.head(n=10)

bldg_contract_town      Region   Area Population Density
0          ANG MO KIO  North-East  13.94    165,710  12,000
1               BEDOK        East  21.69    281,300  13,000
2              BISHAN     Central   7.62     88,490  12,000
3            BOON LAY        West   8.23         30     3.6
4         BUKIT BATOK        West  11.13    144,410  13,000
5         BUKIT MERAH     Central  14.34    151,870  11,000
6       BUKIT PANJANG        West   8.99    140,820  16,000
7         BUKIT TIMAH     Central  17.53     77,280   4,400
9              CHANGI        East  40.61      2,080    62.3
11      CHOA CHU KANG        West   6.11    187,510  31,000

<h2 style='text-decoration:underline'>Number of Dwellings Sold by Region</h2>

We want to analyze which region of Singapore has the most number of dwelling units sold 

We merge the two datasets that we have obtained before to obtain the full information on al the building area clusters

In [76]:
hdb_prop_full_df = pd.merge(hdb_prop_gb, planning_areas_df, on='bldg_contract_town')

hdb_prop_full_df.head()

bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
0        BUKIT TIMAH                  2554           0          49   
1      MARINE PARADE                  7860           0          26   
2             BISHAN                 20072           0          17   
3          SERANGOON                 21634           0         161   
4        JURONG EAST                 24122           0         323   

   3room_sold  4room_sold  5room_sold  exec_sold      Region   Area  \
0         439         920         683        380     Central  17.53   
1        3032        1798        1679          0     Central   6.12   
2        2357        9359        5716       1567     Central   7.62   
3        4535       10231        3757       2365  North-East   10.1   
4        7046        8188        5925       1871        West  17.83   

  Population   Density  
0     77,280     4,400  
1     47,250     7,700  
2     88,490    12,000  
3    117,310  11,945.2  
4     81,180     4,600

Find the rows that is being removed after merging the two dataframes

In [77]:
print("Length of hdb_prop_gb:", len(hdb_prop_gb))
print("Length of hdb_full_df:", len(hdb_prop_full_df))

df = pd.merge(hdb_prop_gb, hdb_prop_full_df, how='outer', indicator=True)

rows_in_df1_not_in_df2 = df[df['_merge']=='left_only'][hdb_prop_gb.columns]

rows_in_df1_not_in_df2.head()

Length of hdb_prop_gb: 26
Length of hdb_full_df: 25


bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
2       CENTRAL AREA                 12003           0         345   

   3room_sold  4room_sold  5room_sold  exec_sold  
2        4214        3424         902          6

**Central Area is not included** in the merged dataframe that we created as it is not listed in the planning areas of Singapore's data.
As such, we have to manually add it into the merged dataframe so that we can have a complete statistic of all the regions of Singapore

In [78]:
hdb_prop_full_added_df = hdb_prop_full_df.append(rows_in_df1_not_in_df2, ignore_index = False, sort=False)

mask = hdb_prop_full_added_df['bldg_contract_town'] == "CENTRAL AREA"

hdb_prop_full_added_df.loc[mask, 'Region'] = 'Central'

# Uncomment to verify region has been added to CENTRAL AREA
hdb_prop_full_added_df.loc[hdb_prop_full_added_df['bldg_contract_town'] == "CENTRAL AREA"]


bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
2       CENTRAL AREA                 12003           0         345   

   3room_sold  4room_sold  5room_sold  exec_sold   Region Area Population  \
2        4214        3424         902          6  Central  NaN        NaN   

  Density  
2     NaN

It is okay for the Central Area to have NaN values in the Area, Population and Density as we will not be using those values for this portion of the analysis

We will gauge which **Region of Singapore** has the **highest density of the dwelling units** sold by creating a Bubble Map based on the **location of these Regions**.

In [79]:
#Create the function to match prefix -> Latitude and Longitude
def get_longitude(prefix):
    for index, row in bld_area_cluster_names_df.iterrows():
        if(prefix == row[1].strip()):
            return row[3]
        
def get_latitude(prefix):
    for index, row in bld_area_cluster_names_df.iterrows():
        if(prefix == row[1].strip()):
            return row[2]

In [80]:
# Add the Longitude and Latitude to the Dataframe
hdb_prop_full_added_df['longitude'] = hdb_prop_full_added_df['bldg_contract_town'].apply(get_longitude)

hdb_prop_full_added_df['latitude'] = hdb_prop_full_added_df['bldg_contract_town'].apply(get_latitude)

hdb_prop_full_added_df.head()

bldg_contract_town  total_dwelling_units  1room_sold  2room_sold  \
0        BUKIT TIMAH                  2554           0          49   
1      MARINE PARADE                  7860           0          26   
2             BISHAN                 20072           0          17   
3          SERANGOON                 21634           0         161   
4        JURONG EAST                 24122           0         323   

   3room_sold  4room_sold  5room_sold  exec_sold      Region   Area  \
0         439         920         683        380     Central  17.53   
1        3032        1798        1679          0     Central   6.12   
2        2357        9359        5716       1567     Central   7.62   
3        4535       10231        3757       2365  North-East   10.1   
4        7046        8188        5925       1871        West  17.83   

  Population   Density    longitude   latitude  
0     77,280     4,400 103.79720000 1.33120000  
1     47,250     7,700 103.90739520 1.30268890  
2     88,490    12,000 103.84824960 1.35145210  
3    117,310  11,945.2 103.87446170 1.36323600  
4     81,180     4,600 103.74229680 1.33311500

We group the building area clusters with respect to the region that they are in. <br>
We use the mean of the latitude and longitude of the building clusters within the same region to find the most centralized location of that region.

In [91]:
# We use the mean of latitude and longitude to find the most centralized location of the region based on the building area clusters within the region
hdb_prop_region_gb = hdb_prop_full_added_df.groupby(['Region'], as_index=False).agg({"total_dwelling_units": "sum",
                                                                                     "latitude":"mean",
                                                                                     "longitude": "mean"})


We plot a Heat Map of the **locations of the building area clusters** to get a rough gauge of **which area the dwelling units sold will be located at.**

In [112]:
heat_map =  fol.Map(location=[1.3521, 103.8198],
                    zoom_start = 10,
                    tiles = "Stamen Terrain")

heat_data = [[row['latitude'],row['longitude']] for index, row in hdb_prop_full_added_df.iterrows()]

# Plot it on the map
HeatMap(heat_data).add_to(heat_map)

heat_map


Create the Bubble Map of **Total Dwelling Units sold by Region** to analyze which region has of Singapore has the **most number of dwelling units sold.**

In [111]:
bubble_map =  fol.Map(location=[1.3521, 103.8198],
                    zoom_start = 11)
color_scale = np.array(['#fddbc7','#f4a582','#d6604d','#b2182b','#67001f'])
count = 0
for each in hdb_prop_region_gb[0:len(hdb_prop_region_gb)].iterrows():
    fol.Circle(
        location = [each[1]['latitude'],each[1]['longitude']], 
        popup = 'Region: ' + '<b>' + each[1]['Region'].upper() + '</b>''<br>''Units Sold:''<b>' + str(each[1]['total_dwelling_units']) + '</b>',
        radius = float(each[1]['total_dwelling_units']/45),
        color = color_scale[count],
        fill_opacity = 0.7,
        fill_color = color_scale[count],
    ).add_to(bubble_map)
    
    count += 1

bubble_map


*(Click on the Circles to view Information about each Region)*
<h3><u>ANALYSIS</u></h3>
From the Bubble Map above, we can roughly gauge in which region are most of the dwelling units sold in Singapore clustered around. 

The **North-East Region** has the **largest cluster of dwelling units** sold.<br>
The **North and East Region** has the **least cluster of dwelling units** sold. 

Comparing the bubble map to the heat map that we plot previously, we can tell that there is a **linear relationship between the number of building area clusters within a region and the total number of dwelling units sold for that region**.

<h2 style='text-decoration:underline'>Total Number of Dwellings Sold by Building Area Clusters</h2>

Now, we will dive even deeper and analyze the number of dwelling units sold based on specific building area clusters

In [35]:
data = [
    go.Bar(
        x = hdb_prop_gb["bldg_contract_town"],
        y = hdb_prop_gb['total_dwelling_units'],
        showlegend = False,
        name = ""
    )
]

layout = dict(
        title = "Total Dwelling Units Sold by Building Cluster",
        xaxis = dict(title ="Building Cluster",automargin=True),
        yaxis = dict(title = "Total Dwelling Units"),
        autosize=True
)

fig = dict(data=data, layout=layout)

cf.iplot(fig)

<h3><u>ANALYSIS</u></h3> 
From the plot, we can see that Jurong West has the most number of residential HDB dwellings (75208) sold (Inclusive of all Unit Types). Bukit Timah has the least HDB Dwellings (2554) sold which is not surprising as most of the residential estates at Bukit Timah are private housing.

<h2 style='text-decoration:underline'>Number of Dwellings Sold By Type for each Building Area Cluster</h2>

Now, We will separate these total dwelling units sold into their respective unit types for further analysis

In [36]:
data = [
    go.Bar(
        x = hdb_prop_gb["bldg_contract_town"],
        y = hdb_prop_gb['1room_sold'],
        showlegend = True,
        name = "1-Room Sold"
    ),
    go.Bar(
        x= hdb_prop_gb['bldg_contract_town'],
        y= hdb_prop_gb['2room_sold'],
        showlegend = True,
        name = "2-Room Sold"
        
    ),
    go.Bar(
        x= hdb_prop_gb['bldg_contract_town'],
        y= hdb_prop_gb['3room_sold'],
        name = "3-Room Sold"
    )
    ,
    go.Bar(
        x= hdb_prop_gb['bldg_contract_town'],
        y= hdb_prop_gb['4room_sold'],
        name = "4-Room Sold"
    )
    ,
    go.Bar(
        x= hdb_prop_gb['bldg_contract_town'],
        y= hdb_prop_gb['5room_sold'],
        name = "5-Room Sold"
    )
    ,
    go.Bar(
        x= hdb_prop_gb['bldg_contract_town'],
        y= hdb_prop_gb['exec_sold'],
        name = "Executive Sold"
    )
]

layout = dict(
        barmode = 'stack',
        title = "Total Dwelling Units Type Sold by Building Cluster",
        xaxis = dict(title ="Building Cluster",automargin=True),
        yaxis = dict(title = "Total Dwelling Units"),
        autosize=True
)

fig = dict(data=data, layout=layout)

cf.iplot(fig)

*(Click on the legend to isolate each trace)*
<h3><u>ANALYSIS</u></h3> 
From the plot above, we can see that although Jurong East has the most number of dwelling units sold, it do not have the most number of dwelling units by type sold.

The only Building Area Cluster with **1-Room Flat sold (255)** is **Bukit Merah**. 1 Room Flat are typically lived in by the elderly or migrant workers and thus we can say that Bukit Merah may have the most number of Elderly and Migrant Workers in the Bukit Merah Cluster.

The Building Area Cluster  with the most **2-Room Flat Sold** is **Bukit Batok (2210)**. 
For **3 Room Flat Sold (24508)** is **Ang Mo Kio**, **4-Room Flat Sold (31324)** is **Seng Kang**,  **5-Room Flat Sold (24144)** is also **Seng Kang** and **Executive Flat Sold (7460)** is Pasir Ris.
